Replace RFC 8941 with RFC 9651 as crate's default version

apasel422 opened the pull request:
1. `GenericBareItem` has two additional variants: `Date` and `DisplayString`.
2. Parsing accepts a `Version` enum that controls whether dates and display strings are permitted. The newer RFC 9651 is the default.
3. Parsing always propagates errors based on the version.
4. Serialization always operates under RFC 9651. It is the API consumer's responsibility to avoid serialization of `Date` and `DisplayString` when using RFC 8941.

Fixes #119

apasel422 commented:
@valenting Do you agree with the overall approach here, or do you have any seconds thoughts versus your comment at https://github.com/undef1nd/sfv/issues/119#issuecomment-2642231695?

valenting commented:
This looks very good.
Some notes before doing a full review:
- I'd like some documentation that explains when each variant should be used. In general it's each HTTP header RFC that specifies which SFV variant should be used, so we should try to explain that briefly.
- Not sure if panicking when serializing is the best call - libraries shouldn't generally panic, but errors might be ignored by crates that use it... tough call.
- I'm not sure if Format is the best name. I'd prefer Variant, or something else that's not loaded. Maybe we could even say `RFC` 🤷

Following this PR, do you have other changes planned, or should we do a 1.0 release soon?

apasel422 commented:
> * I'd like some documentation that explains when each variant should be used. In general it's each HTTP header RFC that specifies which SFV variant should be used, so we should try to explain that briefly.

Will do.

> * Not sure if panicking when serializing is the best call - libraries shouldn't generally panic, but errors might be ignored by crates that use it... tough call.

I'll upload a commit to this PR showing what the non-panicking version of the code looks like.

> * I'm not sure if Format is the best name. I'd prefer Variant, or something else that's not loaded. Maybe we could even say `RFC` 🤷

I'll look into options here.

> Following this PR, do you have other changes planned, or should we do a 1.0 release soon?

I don't have other significant changes planned, but we should probably do a major version bump before version 1.0, as we've accumulated some significant changes since the last release, and we may want to get feedback before committing to 1.0.

apasel422 commented:
> I'll upload a commit to this PR showing what the non-panicking version of the code looks like.

See 55b5d10 for this. I do think the error-propagation is a bit noisy for code that cannot fail in the crate's default mode (RFC 9651). Do you think it would be worth introducing separate `BareItem` enums for the two formats instead? Then all serialization could remain infallible, with the `Format` enum being replaced with a type-level construct.

apasel422 commented:
I'll upload a commit showing what the separate types would look like.

apasel422 commented:
Yet another option would be to allow the serializer to serialize any bare item type (i.e. remain infallible) and not accept a `Format` parameter at all, since it's up to the user code to select the proper format anyway.

valenting commented:
> Yet another option would be to allow the serializer to serialize any bare item type (i.e. remain infallible) and not accept a `Format` parameter at all, since it's up to the user code to select the proper format anyway.

That would be the other side of the choice spectrum from panicking 🙂 But you bring up a great point.
To be honest, I'm not sure how libraries might end up using this API - I doubt that the serializer would be used to serialize arbitrary or user-controlled items, so this might be OK.
Let's go with this one for now. No panicking or errors. The consumer of this API is responsible of determining if Dates and Display Strings are OK to serialize in any specific header. We might reconsider if anyone presents a different use case.

valenting commented:
> > * I'm not sure if Format is the best name. I'd prefer Variant, or something else that's not loaded. Maybe we could even say `RFC` 🤷
> 
> Thinking about this some more, I'd rather not use `RFC`, just in case this crate ever supports a format that doesn't correspond to an RFC specifically. What do you think about `Version`?

Version seems OK.

apasel422 commented:
> Version seems OK.

Done.
